api/v1/storage_autoscaler.go

```diff
@@ -0,0 +1,128 @@
+)
+
+// TimestampedError is the error message with the time stamp.
+type TimestampedError struct {
+	// Message is the error message in case the storage scaling operation has failed.
+	Message string `json:"message,omitempty"`
+	// Timestamp is the time stamp when the error occurred.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+// StorageAutoScalerStatus defines the observed state of StorageAutoScaler
+type StorageAutoScalerStatus struct {
+	// Phase describes the Phase of StorageAutoScaler
+	// +kubebuilder:validation:Enum="";'NotStarted;InProgress;Failed;Succeeded
+	Phase StorageAutoScalerPhase `json:"phase,omitempty"`
+	// Error is the error message in case the storage scaling operation has failed.
+	// +optional
+	// +nullable
+	Error *TimestampedError `json:"error,omitempty"`
+	// LastExpansionStartTime is the time stamp of the last run start of the storage scaling
+	// +optional
+	LastExpansionStartTime metav1.Time `json:"lastExpansionStartTime,omitempty"`
+	// LastExpansionCompletionTime is the time stamp of the last run completion of the storage scaling
+	// +optional
+	LastExpansionCompletionTime metav1.Time `json:"lastExpansionCompletionTime,omitempty"`
+	// The start OSD size is the original size of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD size.
+	// Used for vertical scaling of OSDs.
+	// +optional
+	StartOsdSize resource.Quantity `json:"startOsdSize,omitempty"`
+	// The ExpectedOsdSize is the size that the auto-expansion has decided to set.
+	// This will be set on the storageCLuster CR as the desired size of the OSDs.
+	// Used for vertical scaling of OSDs.
+	// +optional
+	ExpectedOsdSize resource.Quantity `json:"expectedOsdSize,omitempty"`
+	// The start OSD count is the original count of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD count.
+	// Used for horizontal scaling of OSDs.
+	// +optional
+	StartOsdCount uint16 `json:"startOsdCount,omitempty"`
+	// The Expected OSD count is the count that the auto-expansion has decided to set.
+	// This will be set on the storageCluster CR as the desired count of the OSDs.
+	// Used for horizontal scaling of OSDs.
+	// +optional
+	ExpectedOsdCount uint16 `json:"expectedOsdCount,omitempty"`
+	// The start storage capacity is the original storage capacity of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected storage capacity.
+	// +optional
+	StartStorageCapacity resource.Quantity `json:"startStorageCapacity,omitempty"`
+	// The Expected storage capacity is the storage capacity that the auto-expansion has decided to set.
+	// This will be set on the storageCluster CR as the desired storage capacity of the OSDs.
+	// +optional
+	ExpectedStorageCapacity resource.Quantity `json:"expectedStorageCapacity,omitempty"`
```

Comment:
I'm curious what benefit this gives us over `StartOsdSize` and `ExpectedOsdSize`. I think those are the things that actually control whether expansion will continue or not right?

Doesn't this have room for calculation errors between Ceph's reported total capacity and the PV/PVC capacity?

Reply:
we have to use this values  the alerts, https://github.com/red-hat-storage/ocs-operator/pull/3045#discussion_r1985442114